### PR TITLE
Deprecate `convert_dtype` in `Series.apply`

### PR DESCRIPTION
**Pull request by j-bennet**
* Xref https://github.com/pandas-dev/pandas/pull/52257.

Fixes the following upstream failures:

```
dask/dataframe/tests/test_dataframe.py::test_apply: FutureWarning: the convert_dtype parameter is deprecated and will be removed in a future version.  Do ``ser.astype(object).apply()`` instead if you want ``convert_dtype=False``.
dask/dataframe/tests/test_dataframe.py::test_apply_infer_columns: FutureWarning: the convert_dtype parameter is deprecated and will be removed in a future version.  Do ``ser.astype(object).apply()`` instead if you want ``convert_dtype=False``.
```

- [x] Closes https://github.com/dask/dask/issues/10089
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by j-bennet** [comment]
Test failure is legit, fixing.

**Comment by j-bennet** [comment]
Pretty sure test failures are unrelated.

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -6789,8 +6788,10 @@ def _emulate(func, *args, udf=False, **kwargs):
     Apply a function using args / kwargs. If arguments contain dd.DataFrame /
     dd.Series, using internal cache (``_meta``) for calculation
     """
-    with raise_on_meta_error(funcname(func), udf=udf), check_numeric_only_deprecation():
-        return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
+    with check_numeric_only_deprecation():
+        with check_convert_dtype_deprecation():
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:6792]: Can we pull this out and move it higher up in the callstack so it's closer to where `convert_dtype` is actually being specified? This seems too deep down in internals, lots of the `dask.dataframe` codebase uses `_emulate`

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -7003,7 +7004,8 @@ def apply_and_enforce(*args, **kwargs):
     Ensures the output has the same columns, even if empty."""
     func = kwargs.pop("_func")
     meta = kwargs.pop("_meta")
-    df = func(*args, **kwargs)
+    with check_convert_dtype_deprecation():
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:7007]: Similar comment here. Can we move this closer to the source?

File: dask/dataframe/tests/test_dataframe.py [hunk 3]
```diff
@@ -3160,6 +3166,24 @@ def test_apply():
...
+    ddf = dd.from_pandas(df, npartitions=2)
+    if PANDAS_GT_210:
+        ctx = pytest.warns(FutureWarning, match="the convert_dtype parameter")
+        with ctx:
+            expected = df.x.apply(lambda x: x + 1, convert_dtype=convert_dtype)
+        with ctx:
+            result = ddf.x.apply(lambda x: x + 1, convert_dtype=convert_dtype)
+    else:
+        expected = df.x.apply(lambda x: x + 1, convert_dtype=convert_dtype)
+        with pytest.warns(UserWarning, match="You did not provide metadata"):
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:3182]: This is unrelated to the `convert_dtype` deprecation we're primarily checking here. Let's just explicitly provide a `meta` so we don't need to worry about the (unrelated) meta warning

File: dask/dataframe/tests/test_dataframe.py [hunk 4]
```diff
@@ -3160,6 +3166,24 @@ def test_apply():
         ddf.apply(lambda xy: xy, axis="index")
 
 
+@pytest.mark.parametrize("convert_dtype", [True, False])
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:3169]: Do we test `convert_dtype` anywhere else? If not, it'd be good to test that no warning is emitted by default

**Review by jrbourbeau — COMMENTED** [review]
Thanks @j-bennet! Just a few small comments -- overall this looks good

> Pretty sure test failures are unrelated.

Definitely. Merging `main` should fix those.

**Comment by j-bennet** [comment]
@jrbourbeau I addressed all the comments, great tip about `methods.py`, works perfectly. CI is hanging on macos builds, probably transient.

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -4328,21 +4328,19 @@ def apply(self, func, convert_dtype=True, meta=no_default, args=(), **kwds):
...
         if meta is no_default:
             meta = _emulate(
-                M.apply,
-                self._meta_nonempty,
-                func,
-                convert_dtype=convert_dtype,
-                args=args,
-                udf=True,
-                **kwds,
+                M.apply, self._meta_nonempty, func, args=args, udf=True, **kwds
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:4339]: I'm surprised this doesn't need a ` with check_convert_dtype_deprecation()`. Maybe we're emitting multiple, duplicate warnings?

File: dask/dataframe/tests/test_dataframe.py [hunk 6]
```diff
@@ -3160,6 +3166,23 @@ def test_apply():
...
+@pytest.mark.parametrize("convert_dtype", [None, True, False])
+def test_apply_convert_dtype(convert_dtype):
+    """Make sure that explicit convert_dtype raises a warning with pandas>=2.1"""
+    df = pd.DataFrame({"x": [1, 2, 3, 4], "y": [10, 20, 30, 40]})
+    ddf = dd.from_pandas(df, npartitions=2)
+    kwargs = {} if convert_dtype is None else {"convert_dtype": convert_dtype}
+    if PANDAS_GT_210 and convert_dtype is not None:
+        ctx = pytest.warns(FutureWarning, match="the convert_dtype parameter")
+    else:
+        ctx = contextlib.nullcontext()
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:3178]: (Not a block comment, feel free to ignore) this would be a good place to utilize this little `_check_warning` utility 

https://github.com/dask/dask/blob/b86786e3e1bd5be37a24d5b9743c6c043b9daffc/dask/utils_test.py#L154-L164

File: dask/dataframe/core.py (hunk 5 above)
Comment (j-bennet) [reply dask/dataframe/core.py:4339]: Good point, I'll see how many we're emitting.

Comment (j-bennet) [reply dask/dataframe/core.py]: Yes, it was two, only in case there's no meta, which is why I missed it. Fixed now.

**Review by jrbourbeau — APPROVED** [review]
Thanks @j-bennet -- will merge once CI is finished
